feat($animate): add support for `customFilter`

gkalpak opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Feature.

**What is the new behavior (if this is a feature change)?**
This commit adds a new `customFilter()` function on `$animateProvider` (similar to `classNameFilter()`), which can be used to filter animations (i.e. decide whether they are allowed or not), based on the return value of a custom filter function.
This allows to easily create arbitrarily complex rules for filtering animations, such as allowing specific events only, or enabling animations on specific subtrees of the DOM etc.

**Does this PR introduce a breaking change?**
No.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:
Fixes #14891.

Needs docs:
- [x] Complete the API docs.
- [x] Update the Developer Guide.

gkalpak committed:
feat($animate): add support for `customFilter`

This commit adds a new `customFilter()` function on `$animateProvider` (similar
to `classNameFilter()`), which can be used to filter animations (i.e. decide
whether they are allowed or not), based on the return value of a custom filter
function.
This allows to easily create arbitrarily complex rules for filtering animations,
such as allowing specific events only, or enabling animations on specific
subtrees of the DOM, etc.

Fixes #14891

petebacondarwin commented on test/ngAnimate/animateSpec.js:
```diff
@@ -292,6 +292,129 @@ describe("animations", function() {
+          $animateProvider.customFilter(angular.noop);
+          expect($animateProvider.customFilter()).toEqual(jasmine.any(Function));
+
+          $animateProvider.customFilter({});
```
perhaps we should throw instead of silently ignoring?

gkalpak replied:
I kept this consistent with `classNameFilter` (although it is preferrable to throw obviously).
Throwing would be a BC for `classNameFilter`. Should I fix this one to throw and make `classNameFilter` throw for 1.6 (or leave it as is)?

petebacondarwin replied:
I would change both as you suggest.

gkalpak replied:
I missed that this needed to land in 1.6.0 (due to the BC) :(
I will make only `customFilter` throw for 1.6.x/1.7.x and change `classNameFilter` to also throw for 1.7.x only.

gkalpak replied:
Actually, I just realized that the current implementation allows to "clear" the filter/className by passing a non-Function/non-RegExp argument. So, if we want to throw, we need to still allow some special value (e.g. `null`/`undefined`) for unsetting the filter/className.

petebacondarwin commented:
One minor question - otherwise LGTM

Narretz commented:
I like it. I would add a note though that this going to be called for every attempted animation, so developers should keep the function lean.

gkalpak committed:
perf($animate): do not retrieve `className` unless `classNameFilter` is used

gkalpak committed:
docs(guide/animations): list missing animated directives (and other improvements)

- List missing animation-aware directives.
- Fix/Improve wording/formatting.
- Fix typos.
- Limit lines to 100 chars.

gkalpak commented:
I added docs for `customFilter()` (and large-ish docs commit that face-lifts the Animations Guide - sorry).
I didn't make any change wrt throwing on "invalid" argument types, because of https://github.com/angular/angular.js/pull/14914#r94752256.

Feel free to TAL ☺️

Narretz commented on src/ng/animate.js:
```diff
@@ -233,6 +234,45 @@ var $AnimateProvider = ['$provide', /** @this */ function($provide) {
    * @ngdoc method
+   * @name $animateProvider#customFilter
+   *
+   * @description
```
I would add two things:
- How does this relate to the classFilter, .i.e. what takes precedence, or do both run
- Note that this runs for every `animatable` action, not just for animations that are actually defined, and that the user should keep the function fast

You've added this to the guide, but I think it should be here too

gkalpak replied:
> How does this relate to the classFilter, .i.e. what takes precedence, or do both run

👍 

> Note that this runs for every animatable action, not just for animations that are actually defined, and that the user should keep the function fast

You mean in addition to the alert-box below?

Narretz replied:
If possible, I thought we should make it explicit (in the alert box) that this doesn't only affect the actual animations, but every "action" (class change, element entering). I think this difference is easily forgotten, and people might think "Oh well I don't have many animations anyway".

gkalpak replied:
I see what you mean now. Good point. I will update the API docs and guide.
BTW, I think we should re-order the conditions, because now `customFilter` and `classNameFilter` are executed even if animations are globally disabled or the document is hidden etc.

Narretz commented on docs/content/guide/animations.ngdoc:
```diff
@@ -6,20 +6,25 @@
-triggered, will attempt to perform a CSS Transition, CSS Keyframe Animation or a JavaScript callback Animation (depending on if an animation is
-placed on the given directive). Animations can be placed using vanilla CSS by following the naming conventions set in place by AngularJS
-or with JavaScript code when it's defined as a factory.
+AngularJS provides animation hooks for common directives such as `ngRepeat`, `ngSwitch`, and
```
While you are at it, can you add links to the directives here?

Narretz commented on docs/content/guide/animations.ngdoc:
```diff
@@ -255,17 +259,19 @@ myModule.directive('my-directive', ['$animate', function($animate) {
-This ensures that the whole app has been compiled fully before animations are attempted.
+Internally, `ngAnimate` waits until all template downloads that are started right after bootstrap
+have finished. Then, it waits for the currently running {@link ng.$rootScope.Scope#$digest $digest}
+and the one more after that to finish. This ensures that the whole app has been compiled fully
```
"and one more after that, to finish."

Narretz commented on docs/content/guide/animations.ngdoc:
```diff
@@ -294,42 +324,44 @@ app.config(function($animateProvider) {
 
 When the first argument is a native DOM or jqLite/jQuery element, the function enables / disables
 animations on this element *and all its children*: `$animate.enabled(myElement, false)`. This is the
 most flexible way to change the animation state. For example, even if you have used it to disable
```
Is "This is the most flexible way to change the animation state." still true with the new customFilter? I don't think so, as $animate.enabled() does not work when the classNameFilter or customFilter are active. So let's remove this sentence

Narretz commented on docs/content/guide/animations.ngdoc:
```diff
@@ -379,66 +412,71 @@ In that case, you can add styles to the CSS that make sure the element stays hid
 
 
-### Enable animations for elements outside of the Angular application DOM tree: {@link ng.$animate#pin $animate.pin()}
+### Enable animations outside of the application DOM tree: {@link ng.$animate#pin $animate.pin()}
```
This should be `##`

gkalpak commented:
Comments addressed. PTAL

gkalpak commented:
Note to self: Update links after #15654.

Narretz commented:
This looks (almost) done, wanna put it into 1.6.5?

Narretz commented:
@gkalpak Looks like this needs a rebase. After that, I will do another review - or do you have some other changes to make?

gkalpak committed:
perf(animate): avoid unnecessary computations if animations are globally disabled

gkalpak committed:
fixup! feat($animate): add support for `customFilter`

gkalpak commented:
I have squashed and rebased this and added two new commits:
- One for further clarifying when the `customFilter` function is called.
- One for skipping the `customFilter` and `classNameFilter` checks, if animations are globally disabled.

@Narretz PTAL

Narretz commented on docs/content/guide/animations.ngdoc:
```diff
@@ -59,191 +65,190 @@ You may also want to setup a separate CSS file for defining CSS-based animations
 With these generated CSS class names present on the element at the time, AngularJS automatically
-figures out whether to perform a CSS and/or JavaScript animation. If both CSS and JavaScript animation
-code is present, and match the CSS class name on the element, then AngularJS will run both animations at the same time.
+figures out whether to perform a CSS and/or JavaScript animation. If both CSS and JavaScript
```
That actually hasn't been true since 1.4 😱  https://code.angularjs.org/snapshot/docs/guide/migration#animation-nganimate-

gkalpak replied:
So, what happens if both are detected. Who wins?

Narretz replied:
JS animations - because you can run CSS in JS: http://plnkr.co/edit/rdRAZTiXHNAFTEOHyZVo?p=preview

Although I'm surprised this isn't explicitly said in the migrate docs. But it is here: https://code.angularjs.org/snapshot/docs/api/ngAnimate

Unfortuante spreading out of content ... moving this in guide would be a good passion project for rainy afternoons. :p

gkalpak replied:
Changed and added a link to that `api/ngAnimate` section.
There is indeed useful info there that would be better in the guide 😞

Narretz commented on docs/content/guide/animations.ngdoc:
```diff
@@ -59,191 +65,190 @@ You may also want to setup a separate CSS file for defining CSS-based animations
+AngularJS also pays attention to CSS class changes on elements by triggering the **add** and
+**remove** hooks. This means that if a CSS class is added to or removed from an element then an
+animation can be executed in between, before the CSS class addition or removal is finalized.
+(Keep in mind that AngularJS will only be able to capture class changes if an **expression** or the
```
I think **interpolated expression** is clearer here

gkalpak replied:
Fixed.

Narretz reviewed (changes requested):
Just two smaller docs changes needed, otherwise LGTM

gkalpak committed:
fixup! docs(guide/animations): list missing animated directives (and other improvements)
